DOCS-#128: Update installation section in docs (#132)
modin-project/unidist: docs/installation.rst

on the `GitHub repo`_. The master branch has the most recent patches, but may be less
stable than a release installed from ``pip`` or ``conda``.

Installing with pip
'''''''''''''''''''

Stable version
""""""""""""""

unidist can be installed with ``pip`` on Linux, Windows and MacOS.
To install the most recent stable release run the following:

.. code-block:: bash

  pip install unidist # Install unidist dependencies for Multiprocessing and sequential Python backends

unidist can also be used with Dask, MPI or Ray execution backend.
If you don't have Dask_, MPI_ or Ray_ installed, you will need to install unidist with one of the targets:

.. code-block:: bash

  pip install unidist[all] # Install unidist dependencies for all the backends
  pip install unidist[dask] # Install unidist dependencies for Dask backend
  pip install unidist[mpi] # Install unidist dependencies for MPI backend
  pip install unidist[ray] # Install unidist dependencies for Multiprocessing and sequential Python backends

unidist will automatically detect which execution backend you have installed and use that for
scheduling computation!

Release candidates
""""""""""""""""""

> unidist follows `Semantic Versioning`_ and before some major or minor releases,

Reviewer: Could you please explain more about pre-release builds. It is a part of general release processes but we decide if we add it to conda or not?

Author: We may want to upload pre-release builds on PyPI before some major or minor releases at some point in order to test and check if there are any problems. This is not related to conda because we first upload packages on PyPI.

+---------------------------------+-----------------------------------------------------+-----------------------------+
| unidist-all                     | `Dask`_, `MPI`_, `Ray`_, `Multiprocessing`_, Python | Linux, Windows              |
+---------------------------------+-----------------------------------------------------+-----------------------------+
| unidist-dask                    | `Dask`_                                             | Linux, Windows, MacOS       |
+---------------------------------+-----------------------------------------------------+-----------------------------+
| unidist-mpi                     | `MPI`_                                              | Linux, Windows, MacOS       |
+---------------------------------+-----------------------------------------------------+-----------------------------+
| unidist-ray                     | `Ray`_                                              | Linux, Windows              |
+---------------------------------+-----------------------------------------------------+-----------------------------+

> Installing unidist packages from the conda-forge channel can be achieved by adding conda-forge to your channels with:

Reviewer: Worth mentioning how to add forge channel?

.. code-block:: bash

  conda config --add channels conda-forge
  conda config --set channel_priority strict

For installing unidist with Dask and MPI execution backends into a conda environment the following command should be used:

.. code-block:: bash

  conda install unidist-dask unidist-mpi -c conda-forge

All set of backends could be available in a conda environment by specifying:

.. code-block:: bash

  conda install unidist-all -c conda-forge

or explicitly:

.. code-block:: bash

  conda install unidist-dask unidist-mpi unidist-ray -c conda-forge

Installing from the GitHub master branch
''''''''''''''''''''''''''''''''''''''''

If you'd like to try unidist using the most recent updates from the master branch, you can
also use ``pip``.

.. code-block:: bash

  pip install git+https://github.com/modin-project/unidist

This will install directly from the repo without you having to manually clone it! Please be aware
that these changes have not made it into a release and may not be completely stable.

Building unidist from Source
''''''''''''''''''''''''''''

> If you're planning on :doc:`contributing </developer/contributing>` to unidist, you will need to ensure that you are

Reviewer: ```suggestion
If you're planning to :doc:`contribute </developer/contributing>`, you need to ensure that you are
```

Author: corrected

> building unidist from the local repository that you are working off of. Occasionally,

Reviewer: ```suggestion
building unidist from the local repository that you are working of. Occasionally,
```

Author: fixed

there are issues in overlapping unidist installs from PyPI and from source. To avoid these

> issues, we recommend uninstalling unidist before you install from source:

Reviewer: ```suggestion
issues, we recommend uninstalling unidist before installation from source:
```

``pip`` or using ``conda`` tool, but some users may want to build from the master branch
on the `GitHub repo`_. The master branch has the most recent patches, but may be less
stable than a release installed from ``pip`` or ``conda``.

Installing with pip
'''''''''''''''''''

Stable version
""""""""""""""

unidist can be installed with ``pip`` on Linux, Windows and MacOS.
To install the most recent stable release run the following:

.. code-block:: bash

  pip install unidist # Install unidist with dependencies for Multiprocessing and sequential Python backends

unidist can also be used with Dask, MPI or Ray execution backend.
If you don't have Dask_, MPI_ or Ray_ installed, you will need to install unidist with one of the targets:

.. code-block:: bash

  pip install unidist[all] # Install unidist with dependencies for all the backends
  pip install unidist[dask] # Install unidist with dependencies for Dask backend
  pip install unidist[mpi] # Install unidist with dependencies for MPI backend
  pip install unidist[ray] # Install unidist with dependencies for Multiprocessing and sequential Python backends

> unidist automatically detects which execution backends are installed and use that for

Author: corrected

| unidist-all                     | `Dask`_, `MPI`_, `Ray`_, `Multiprocessing`_, Python | Linux, Windows              |
+---------------------------------+-----------------------------------------------------+-----------------------------+
| unidist-dask                    | `Dask`_                                             | Linux, Windows, MacOS       |
+---------------------------------+-----------------------------------------------------+-----------------------------+
| unidist-mpi                     | `MPI`_                                              | Linux, Windows, MacOS       |
+---------------------------------+-----------------------------------------------------+-----------------------------+
| unidist-ray                     | `Ray`_                                              | Linux, Windows              |
+---------------------------------+-----------------------------------------------------+-----------------------------+

Installing unidist packages from the conda-forge channel can be achieved by adding conda-forge to your channels with:

.. code-block:: bash

  conda config --add channels conda-forge

>   conda config --set channel_priority strict

Reviewer 2: I'm not sure that this is needed actions. We already specify channel `conda-forge` during installation. Why do I need to change my conda config?

Author: Some inattentive users may miss `-c conda-forge` in the installation line so I think we should be explicit.

Author: ok, let's remove for now

For installing unidist with dependencies for Dask and MPI execution backends into a conda environment
the following command should be used:

.. code-block:: bash

  conda install unidist-dask unidist-mpi -c conda-forge

All set of backends could be available in a conda environment by specifying:

.. code-block:: bash

  conda install unidist-all -c conda-forge

or explicitly:

.. code-block:: bash

  conda install unidist-dask unidist-mpi unidist-ray -c conda-forge

Installing from the GitHub master branch
''''''''''''''''''''''''''''''''''''''''

If you'd like to try unidist using the most recent updates from the master branch, you can
also use ``pip``.

.. code-block:: bash

>   pip install git+https://github.com/modin-project/unidist

Reviewer 2: I suggest to add additional example of installing from sources with some set of dependencies ([ray]/[dask] etc).

Author: Good catch!

This will install directly from the repo without you having to manually clone it! Please be aware
that these changes have not made it into a release and may not be completely stable.

Building unidist from Source
''''''''''''''''''''''''''''

If you're planning to :doc:`contribute </developer/contributing>` to unidist, you need to ensure that you are
building unidist from the local repository that you are working of. Occasionally,
there are issues in overlapping unidist installs from PyPI and from source. To avoid these
issues, we recommend uninstalling unidist before installation from source:

.. code-block:: bash

  pip uninstall unidist

To build from source, you first must clone the repo. We recommend forking the repository first
through the GitHub interface, then cloning as follows:

.. code-block:: bash

  git clone https://github.com/<your-github-username>/unidist.git

Once cloned, ``cd`` into the ``unidist`` directory and use ``pip`` to install:

.. code-block:: bash

  cd unidist

>   pip install -e .

Reviewer 2: The same comment as above about dependencies sets.
